File: shared/python/aeos_shared/models/workflow.py
"""Workflow, step, and execution plan data models."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class WorkflowStatus(str, Enum):
    """Workflow lifecycle status."""

    planning = "planning"
    executing = "executing"
    suspended = "suspended"
    completed = "completed"
    failed = "failed"


class StepStatus(str, Enum):
    """Individual workflow step status."""

    pending = "pending"
    active = "active"
    completed = "completed"
    failed = "failed"
    suspended = "suspended"


class AgentType(str, Enum):
    """Specialist agent types in the AEOS system."""

    planner = "planner"
    incident_analysis = "incident_analysis"
    operations = "operations"
    compliance = "compliance"
    validation = "validation"
    recovery = "recovery"
    escalation = "escalation"
    memory = "memory"


class ActionDescriptor(BaseModel):
    """Describes a discrete action to be executed by an agent."""

    model_config = ConfigDict(from_attributes=True)

    tool: str
    params: dict
    timeout_seconds: int = Field(default=30)


class WorkflowStep(BaseModel):
    """A single step within a workflow execution plan."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workflow_id: UUID
    agent_type: AgentType
    action: ActionDescriptor
    status: StepStatus
    depends_on: list[UUID] = Field(default_factory=list)
    risk_score: float | None = Field(default=None, ge=0.0, le=10.0)
    output: dict | None = None
    retry_count: int = Field(default=0)
    created_at: datetime
    updated_at: datetime


class ExecutionPlan(BaseModel):
    """A complete execution plan composed of ordered, dependency-linked steps."""

    model_config = ConfigDict(from_attributes=True)

    steps: list[WorkflowStep]
    metadata: dict = Field(default_factory=dict)


class Workflow(BaseModel):
    """Represents a workflow orchestrating the resolution of an incident."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    incident_id: UUID
    plan: ExecutionPlan | None = None
    status: WorkflowStatus
    current_step_ids: list[UUID] = Field(default_factory=list)
    retry_count: int = Field(default=0)
    checkpoint: dict | None = None
    created_at: datetime
    updated_at: datetime

File: shared/typescript/types/workflow.ts
/**
 * Workflow, step, and execution plan types.
 * Mirrors the Python models in shared/python/aeos_shared/models/workflow.py
 */

export type WorkflowStatus =
  | 'planning'
  | 'executing'
  | 'suspended'
  | 'completed'
  | 'failed';

export type StepStatus =
  | 'pending'
  | 'active'
  | 'completed'
  | 'failed'
  | 'suspended';

export type AgentType =
  | 'planner'
  | 'incident_analysis'
  | 'operations'
  | 'compliance'
  | 'validation'
  | 'recovery'
  | 'escalation'
  | 'memory';

export interface ActionDescriptor {
  tool: string;
  params: Record<string, unknown>;
  timeout_seconds: number;
}

export interface WorkflowStep {
  id: string;
  workflow_id: string;
  agent_type: AgentType;
  action: ActionDescriptor;
  status: StepStatus;
  depends_on: string[];
  risk_score: number | null;
  output: Record<string, unknown> | null;
  retry_count: number;
  created_at: string;
  updated_at: string;
}

export interface ExecutionPlan {
  steps: WorkflowStep[];
  metadata: Record<string, unknown>;
}

export interface Workflow {
  id: string;
  incident_id: string;
  plan: ExecutionPlan | null;
  status: WorkflowStatus;
  current_step_ids: string[];
  retry_count: number;
  checkpoint: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

/**
 * A node in the dependency graph representing a single workflow step.
 */
export interface DependencyGraphNode {
  step_id: string;
  agent_type: AgentType;
  status: StepStatus;
  depends_on: string[];
  /** Step IDs that depend on this node */
  dependents: string[];
}

/**
 * Live dependency graph maintained by the Coordinator per active workflow.
 * Mirrors the in-memory DAG stored in Redis.
 * Satisfies Requirement 4.3.
 */
export interface DependencyGraph {
  workflow_id: string;
  nodes: Record<string, DependencyGraphNode>;
  /** Topologically sorted step IDs (root steps first) */
  execution_order: string[];
}
